WordPress/Classes/Services/ReaderTopicService+Subscriptions.swift

```diff
@@ -0,0 +1,213 @@
+                                                                             into: managedObjectContext) as? ReaderSiteInfoSubscriptionPost
+        }
+
+        if siteTopic.emailSubscription == nil {
+            siteTopic.emailSubscription = NSEntityDescription.insertNewObject(forEntityName: ReaderSiteInfoSubscriptionEmail.classNameWithoutNamespaces(),
+                                                                              into: managedObjectContext) as? ReaderSiteInfoSubscriptionEmail
+        }
+
+        ContextManager.sharedInstance().saveContextAndWait(managedObjectContext)
+
+        return siteTopic
+    }
+
+    private func remoteAction(for action: SubscriptionAction, _ subscribe: Bool, _ success: @escaping () -> Void, _ failure: @escaping (NSError?) -> Void) {
+        let service = ReaderTopicServiceRemote(wordPressComRestApi: apiRequest())
+
+        let successBlock = {
+            ContextManager.sharedInstance().save(self.managedObjectContext, withCompletionBlock: success)
+        }
+
+        switch action {
+        case .notifications(let siteId):
+            if subscribe {
+                service.subscribeSiteNotifications(with: siteId, successBlock, failure)
+            } else {
+                service.unsubscribeSiteNotifications(with: siteId, successBlock, failure)
+            }
+
+        case .postsEmail(let siteId):
+            if subscribe {
+                service.subscribePostsEmail(with: siteId, successBlock, failure)
+            } else {
+                service.unsubscribePostsEmail(with: siteId, successBlock, failure)
+            }
+
+        case .updatePostsEmail(let siteId, let frequency):
+            service.updateFrequencyPostsEmail(with: siteId, frequency: frequency, successBlock, failure)
+
+        case .comments(let siteId):
+            if subscribe {
+                service.subscribeSiteComments(with: siteId, successBlock, failure)
+            } else {
+                service.unsubscribeSiteComments(with: siteId, successBlock, failure)
+            }
+        }
+    }
+}
+
+
+extension ReaderTopicService {
+    @nonobjc public func subscribeSiteNotifications(with siteId: NSNumber, _ success: @escaping () -> Void, _ failure: @escaping (NSError?) -> Void) {
+        toggleSiteNotifications(with: siteId, subscribe: true, success, failure)
+    }
+
+    @nonobjc public func unsubscribeSiteNotifications(with siteId: NSNumber, _ success: @escaping () -> Void, _ failure: @escaping (NSError?) -> Void) {
+        toggleSiteNotifications(with: siteId, success, failure)
+    }
+
+
+    // MARK: Private methods
+
+    private func toggleSiteNotifications(with siteId: NSNumber, subscribe: Bool = false, _ success: @escaping () -> Void, _ failure: @escaping (NSError?) -> Void) {
+        guard let siteTopic = fetchSiteTopic(with: siteId, failure) else {
+            return
+        }
+
+        let oldValue = !subscribe
```

Comment:
Is it possible that this method could ever be called twice in a row with the same `subscribe` value? Perhaps we should set `oldValue` to `siteTopic.postSubscription?.sendPosts` instead? Otherwise, there's an edge case where it could be called with the same `subscribe` value as the current setting, which we would then negate if the request fails? (same with the other methods in here)

Reply:
👍